tensorcircuit/mpscircuit.py

```diff
@@ -335,42 +597,42 @@ def is_valid(self) -> bool:
         return True
 
     @staticmethod
-    def from_wavefunction(
+    def wavefunction_to_tensors(
         wavefunction: Tensor,
-        max_singular_values: Optional[int] = None,
-        max_truncation_err: Optional[float] = None,
-        relative: bool = True,
-    ) -> "MPSCircuit":
+        dim_phys: int = 2,
+        norm: bool = True,
+        **rules: Any,
+    ) -> List[Tensor]:
         """
-        Construct the MPS from a given wavefunction.
+        Construct the MPS tensors from a given wavefunction.
 
         :param wavefunction: The given wavefunction (any shape is OK)
         :type wavefunction: Tensor
-        :param max_singular_values: The maximum number of singular values to keep.
-        :type max_singular_values: int, optional
-        :param max_truncation_err: The maximum allowed truncation error.
-        :type max_truncation_err: float, optional
-        :param relative: Multiply `max_truncation_err` with the largest singular value.
-        :type relative: bool, optional
-        :return: The constructed MPS
-        :rtype: MPSCircuit
+        :param rules: Truncation rules
+        :type rules: Dict
+        :param dim_phys: Physical dimension, 2 for MPS and 4 for MPO
+        :type dim_phys: int
+        :param norm: Whether to normalize the wavefunction
+        :type norm: bool
+        :return: The tensors
+        :rtype: List[Tensor]
         """
         wavefunction = backend.reshape(wavefunction, (-1, 1))
         tensors: List[Tensor] = []
         while True:  # not jittable
             nright = wavefunction.shape[1]
-            wavefunction = backend.reshape(wavefunction, (-1, nright * 2))
+            wavefunction = backend.reshape(wavefunction, (-1, nright * dim_phys))
             wavefunction, Q = split_tensor(
                 wavefunction,
-                left=True,
-                max_singular_values=max_singular_values,
-                max_truncation_err=max_truncation_err,
-                relative=relative,
+                center_left=True,
+                **rules,
             )
-            tensors.insert(0, backend.reshape(Q, (-1, 2, nright)))
+            tensors.insert(0, backend.reshape(Q, (-1, dim_phys, nright)))
             if wavefunction.shape == (1, 1):
                 break
-        return MPSCircuit(len(tensors), tensors=tensors)
+        if not norm:
+            tensors[-1] *= wavefunction[0, 0]
```

Comment:
not clear about this line, how this guarantee the normalization

Reply:
should be `/sqrt(Tr(tensors[0]@tensors[0]^*))`, is there any further simplification I overlook?

Reply:
The code inside the loop is doing QR and adding Q to tensor (R is the wavefunction here), so the Q list (i.e. the tensors) is automatically normalized, and the final R (i.e. the wavefunction) after the loop is a shape (1, 1) tensor which contains the norm.